// snk-libs/lib/utils/CheckUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright (c) 2014-2017. JarkimZhu
 * This software can not be used privately without permission
 */

/**
 * @author JarkimZhu
 * Created on 2016-06-29.
 * @version 0.1.0
 * @class
 */
var CheckUtils = function () {
  function CheckUtils() {
    (0, _classCallCheck3.default)(this, CheckUtils);
  }

  (0, _createClass3.default)(CheckUtils, null, [{
    key: "checkMobile",
    value: function checkMobile(mobile) {
      return (/^1[3|4|5|7|8]\d{9}$/.test(mobile)
      );
    }
  }, {
    key: "checkEmail",
    value: function checkEmail(email) {
      return (/^(\w)+(\.\w+)*@(\w)+((\.\w+)+)$/.test(email)
      );
    }
  }, {
    key: "checkQQ",
    value: function checkQQ(qq) {
      return (/^\d{5,}$/.test(qq) || CheckUtils.checkEmail(qq)
      );
    }
  }, {
    key: "checkWechat",
    value: function checkWechat(wx) {
      return CheckUtils.checkMobile(wx) || CheckUtils.checkEmail(wx) || CheckUtils.checkQQ(wx) || /^[a-zA-Z\d_]{5,}$/.test(wx);
    }
  }, {
    key: "checkSmsCode",
    value: function checkSmsCode(smsCode) {
      return (/\d{6}/.test(smsCode)
      );
    }
  }, {
    key: "checkIdentityCode",
    value: function checkIdentityCode(identityCode) {
      return (/^(\d{15}$|^\d{18}$|^\d{17}(\d|X|x))$/.test(identityCode)
      );
    }
  }, {
    key: "checkName",
    value: function checkName(name) {
      return (/^[\u0391-\uFFE5A-Za-z]+$/.test(name)
      );
    }
  }, {
    key: "checkPostCode",
    value: function checkPostCode(name) {
      return (/^[1-9][0-9]{5}$/.test(name)
      );
    }
  }]);
  return CheckUtils;
}();

exports.default = CheckUtils;
