// snk-libs/lib/network/HttpClient.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _AjaxHttpClient = require('./AjaxHttpClient');

var _AjaxHttpClient2 = _interopRequireDefault(_AjaxHttpClient);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @version 0.0.1-SNAPSHOT
 * @author JarkimZhu
 * Created on 2016/3/18.
 * @class
 */
var HttpClient = function () {
  function HttpClient() {
    (0, _classCallCheck3.default)(this, HttpClient);
    this._impl = null;

    this._impl = new _AjaxHttpClient2.default(false);
  }

  (0, _createClass3.default)(HttpClient, [{
    key: 'get',
    value: function get(url, request, isScriptGet, enableCache) {
      return this._impl.get(url, request, isScriptGet, enableCache);
    }
  }, {
    key: 'post',
    value: function post(url, request) {
      return this._impl.post(url, request);
    }
  }, {
    key: 'put',
    value: function put(url, request) {
      return this._impl.put(url, request);
    }
  }, {
    key: 'delete',
    value: function _delete(url, request) {
      return this._impl.delete(url, request);
    }
  }, {
    key: 'patch',
    value: function patch(url, request) {
      return this._impl.patch(url, request);
    }
  }, {
    key: 'upload',
    value: function upload(url, request, pcb) {
      return this._impl.upload(url, request, pcb);
    }
  }, {
    key: 'clearCookie',
    value: function clearCookie() {
      this._impl.clearCookie();
    }
  }]);
  return HttpClient;
}(); /*
      * Copyright (c) 2014-2016. JarkimZhu
      * This software can not be used privately without permission
      */

exports.default = HttpClient;
